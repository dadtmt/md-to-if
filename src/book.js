import * as R from 'ramda'
import { snakeCase } from 'change-case'

export const getSceneName = R.pipe(
  R.propOr([], 'content'),
  R.head,
  R.propOr('unnamed', 'content'),
  snakeCase
)

export const splitByScene = R.pipe(
  R.converge(R.prepend, [
    R.head,
    R.pipe(
      R.tail,
      R.splitWhen(
        R.where({
          type: R.equals('heading'),
          level: R.equals(2),
        })
      )
    ),
  ]),
  R.zipObj(['heading', 'content', 'sourceLeft']),
  ({ heading, content, sourceLeft }) => ({
    scene: {
      name: getSceneName(heading),
      content: [heading, ...content],
    },
    sourceLeft,
  })
)

const book = (source, scenes = []) => {
  const { scene, sourceLeft } = splitByScene(source)
  return R.isEmpty(sourceLeft)
    ? [...scenes, scene]
    : book(sourceLeft, [...scenes, scene])
}

export default book
