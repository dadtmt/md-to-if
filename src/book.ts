import * as R from 'ramda'
import { snakeCase } from 'change-case'

// Content -> String
export const getSceneName = R.pipe(
  R.propOr([], 'content'),
  R.head,
  R.propOr('unnamed', 'content'),
  snakeCase
)

// int -> [Content] -> [[Content],[Content]]
const splitByHeading = level =>
  R.splitWhen(
    R.where({
      type: R.equals('heading'),
      level: R.equals(level),
    })
  )

const splitActions = (level, contentList, actionList = []) => {
  const [headOfContent, ...tailOfContent] = contentList
  const [actionContent, restOfContent] = splitByHeading(level)(tailOfContent)
  const { content, actions } =
    level <= 6
      ? splitContentAndActions(level + 1)(actionContent)
      : {
          content: [],
          actions: [],
        }
  return !R.isEmpty(restOfContent)
    ? splitActions(level, restOfContent, [
        ...actionList,
        {
          name: getSceneName(headOfContent),
          sceneContent: [headOfContent, ...content],
          actions,
        },
      ])
    : actionList
}

// [Content] -> { content: [Content], actions: [Scene] }
export const splitContentAndActions = (level: number) => contentAndActions => {
  const [content, actions] = splitByHeading(level)(contentAndActions)
  return { content, actions: splitActions(level, actions) }
}

// [Content] -> { scene: Scene, content: [Content] }
export const splitByScene = (level: number): Function =>
  R.pipe(
    R.converge(R.prepend, [R.head, R.pipe(R.tail, splitByHeading(level))]),
    R.zipObj(['heading', 'content', 'sourceLeft']),
    ({ content, ...rest }) => ({
      content,
      ...rest,
    }),
    ({ heading, content, sourceLeft }) => ({
      scene: {
        name: getSceneName(heading),
        sceneContent: [heading, ...content],
      },
      sourceLeft,
    })
  )

// [Content], [Scene] -> [Scene]
const book = (source, scenes = []) => {
  const { scene, sourceLeft } = splitByScene(2)(source)
  return R.isEmpty(sourceLeft)
    ? [...scenes, scene]
    : book(sourceLeft, [...scenes, scene])
}

export default book
