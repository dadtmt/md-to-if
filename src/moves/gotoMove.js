import * as R from 'ramda'
import { incPlayedSceneCount } from './playedSceneCount'

export const getTarget = R.pipe(R.prop('target'), R.tail)

export const matchTarget = move => R.propEq('name', getTarget(move))

export const gotoScene = move => R.find(matchTarget(move))

const incrementPlayedScene = name =>
  R.pipe(R.assoc('currentSceneName', name), incPlayedSceneCount(name))

const goto = (scenes, story) => [
  R.propEq('type', 'anchor'),
  move => {
    const lastState = R.pipe(R.last, R.prop('state'))(story)
    const scene = gotoScene(move)(scenes)
    const { name } = scene

    return {
      ...scene,
      state: incrementPlayedScene(name)(lastState),
    }
  },
]

export default goto
