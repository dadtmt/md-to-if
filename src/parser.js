import SimpleMarkdown from 'simple-markdown'

// String -> [String]
export const matchBraces = source => /^\{([\s\S]*?)\}/.exec(source)

// String -> [String]
export const matchBracketPipe = source => /^\[([\s\S]*?)\|/.exec(source)

// String -> [String]
export const matchPipeBracket = source => /^\|([\s\S]*?)\]/.exec(source)

const command = {
  order: 0,
  match: matchBraces,
  parse: (capture, parse, state) => ({ content: parse(capture[1], state) }),
}

const trueCaseContent = {
  order: 0,
  match: matchBracketPipe,
  parse: (capture, parse, state) => ({ content: parse(capture[1], state) }),
}

const falseCaseContent = {
  order: 0,
  match: matchPipeBracket,
  parse: (capture, parse, state) => ({ content: parse(capture[1], state) }),
}

// String -> [Content]
const parser = SimpleMarkdown.parserFor({
  ...SimpleMarkdown.defaultRules,
  command,
  trueCaseContent,
  falseCaseContent,
})

export default parser
