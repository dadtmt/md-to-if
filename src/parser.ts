import SimpleMarkdown, { SingleASTNode } from 'simple-markdown'

// String -> [String]
export const matchBraces = (source: string): RegExpExecArray | null =>
  /^\{([\s\S]*?)\}/.exec(source)

// String -> [String]
export const matchBracketPipe = (source: string): RegExpExecArray | null =>
  /^\[([\s\S]*?)\|/.exec(source)

// String -> [String]
export const matchPipeBracket = (source: string): RegExpExecArray | null =>
  /^\|([\s\S]*?)\]/.exec(source)

const parse: (capture: any, parse: any, state: any) => any = (
  capture,
  parse,
  state
) => ({ content: parse(capture[1], state) })

const command = {
  order: 0,
  match: matchBraces,
  parse
}

const trueCaseContent = {
  order: 0,
  match: matchBracketPipe,
  parse
}

const falseCaseContent = {
  order: 0,
  match: matchPipeBracket,
  parse
}

// String -> [Content]
const parser: (source: string) => SingleASTNode[] = SimpleMarkdown.parserFor({
  ...SimpleMarkdown.defaultRules,
  command,
  trueCaseContent,
  falseCaseContent
})

export default parser
