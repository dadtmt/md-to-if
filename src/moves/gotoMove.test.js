import parser from '../parser'
import { gotoScene, getTarget, matchTarget } from './gotoMove'
import adventure from '../adventure.md.js'
import book from '../book'

const adventureBook = book(parser(adventure))

const moveToCantina = {
  type: 'anchor',
  target: '/cantina',
}
const moveToSpaceShip = {
  type: 'anchor',
  target: '/the_space_ship',
}

describe('gotoScene', () => {
  it('goto scene', () => {
    expect(gotoScene(moveToCantina, [])(adventureBook)).toMatchSnapshot()
  })
})

describe('matchTarget', () => {
  it('move to cantina match cantina scene', () => {
    const cantinaScene = adventureBook[2]
    expect(matchTarget(moveToCantina)(cantinaScene)).toBe(true)
  })
  it('move to cantina does not match spaceship scene', () => {
    const spaceShipScene = adventureBook[1]
    expect(matchTarget(moveToCantina)(spaceShipScene)).toBe(false)
  })
  it('move to spaceship match spaceship scene', () => {
    const spaceShipScene = adventureBook[1]
    expect(matchTarget(moveToSpaceShip)(spaceShipScene)).toBe(true)
  })
})

describe('getTarget', () => {
  it('get move target', () => {
    expect(getTarget(moveToCantina)).toBe('cantina')
  })
})
