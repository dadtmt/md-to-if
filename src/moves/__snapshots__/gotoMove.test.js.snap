// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gotoScene goto scene 1`] = `
Object {
  "name": "cantina",
  "sceneContent": Array [
    Object {
      "content": Array [
        Object {
          "content": "Cantina",
          "type": "text",
        },
      ],
      "level": 2,
      "type": "heading",
    },
    Object {
      "content": Array [
        Object {
          "content": " show playedCount ",
          "type": "text",
        },
      ],
      "type": "dynamic",
    },
    Object {
      "content": Array [
        Object {
          "content": " time you come here",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": "The cantina is full with your crew having a good time ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": " test playedCount equals 1 ",
              "type": "text",
            },
          ],
          "type": "dynamic",
        },
        Object {
          "content": " ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": "trueCase",
              "type": "text",
            },
          ],
          "type": "trueCaseContent",
        },
        Object {
          "content": Array [
            Object {
              "content": "falseCase",
              "type": "text",
            },
          ],
          "type": "falseCaseContent",
        },
        Object {
          "content": " ",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": "They are singing ",
          "type": "text",
        },
        Object {
          "content": "'oh my captain",
          "type": "text",
        },
        Object {
          "content": "' for the ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": " show playedCount ",
              "type": "text",
            },
          ],
          "type": "dynamic",
        },
        Object {
          "content": " times",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": "You pick up a sandwich an put it in your bag ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": " set bag sandwich 1 ",
              "type": "text",
            },
          ],
          "type": "dynamic",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": "You have ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": " show bag sandwich ",
              "type": "text",
            },
          ],
          "type": "dynamic",
        },
        Object {
          "content": " with you",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "content": Array [
        Object {
          "content": " test playedCount equals 1 ",
          "type": "text",
        },
      ],
      "type": "dynamic",
    },
    Object {
      "content": Array [
        Object {
          "content": " ",
          "type": "text",
        },
        Object {
          "content": Array [
            Object {
              "content": " This is your first time here
    Will you come back ",
              "type": "text",
            },
            Object {
              "content": "? ",
              "type": "text",
            },
          ],
          "type": "trueCaseContent",
        },
        Object {
          "content": Array [
            Object {
              "content": " Should not you go somewhere else ",
              "type": "text",
            },
            Object {
              "content": "?
or what ",
              "type": "text",
            },
            Object {
              "content": "? ",
              "type": "text",
            },
          ],
          "type": "falseCaseContent",
        },
        Object {
          "content": " ",
          "type": "text",
        },
      ],
      "type": "paragraph",
    },
  ],
}
`;
